sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala

```diff
@@ -551,28 +629,452 @@ class ParquetPartitionReader(
-    if (currentChunkedBlocks.isEmpty) {
-      return None
-    }
-
-    val (dataBuffer, dataSize) = readPartFile(currentChunkedBlocks)
+  protected def dumpParquetData(
+      hmb: HostMemoryBuffer,
+      dataLength: Long,
+      splits: Array[PartitionedFile]): Unit = {
+    val (out, path) = FileUtils.createTempFile(conf, debugDumpPrefix, ".parquet")
     try {
-      if (dataSize == 0) {
-        None
-      } else {
-        if (debugDumpPrefix != null) {
-          dumpParquetData(dataBuffer, dataSize)
-        }
-        val parseOpts = ParquetOptions.builder()
-          .withTimeUnit(DType.TIMESTAMP_MICROSECONDS)
-          .includeColumn(readDataSchema.fieldNames:_*).build()
-
-        // about to start using the GPU
-        GpuSemaphore.acquireIfNecessary(TaskContext.get())
+      logInfo(s"Writing Parquet split data for $splits to $path")
+      val in = new HostMemoryInputStream(hmb, dataLength)
+      IOUtils.copy(in, out)
+    } finally {
+      out.close()
+    }
+  }
+}
 
-        val table = withResource(new NvtxWithMetrics("Parquet decode", NvtxColor.DARK_GREEN,
-            metrics(GPU_DECODE_TIME))) { _ =>
+/**
+ * A PartitionReader that can read multiple Parquet files up to the certain size.
+ *
+ * Efficiently reading a Parquet split on the GPU requires re-constructing the Parquet file
+ * in memory that contains just the column chunks that are needed. This avoids sending
+ * unnecessary data to the GPU and saves GPU memory.
+ *
+ * @param conf the Hadoop configuration
+ * @param split the file split to read
+ * @param clippedBlocks the block metadata from the original Parquet file that has been clipped
+ *                      to only contain the column chunks to be read
+ * @param readDataSchema the Spark schema describing what will be read
+ * @param debugDumpPrefix a path prefix to use for dumping the fabricated Parquet data or null
+ */
+class MultiFileParquetPartitionReader(
+    conf: Configuration,
+    splits: Array[PartitionedFile],
+    clippedBlocks: Seq[MultiFilePartitionInfo],
+    isSchemaCaseSensitive: Boolean,
+    readDataSchema: StructType,
+    debugDumpPrefix: String,
+    maxReadBatchSizeRows: Integer,
+    maxReadBatchSizeBytes: Long,
+    execMetrics: Map[String, SQLMetric],
+    partitionSchema: StructType,
+    isCorrectedRebaseMode: Boolean)
+  extends FileParquetPartitionReaderBase(conf, isSchemaCaseSensitive, readDataSchema,
+    debugDumpPrefix, execMetrics) {
+
+  private val blockIterator: BufferedIterator[MultiFilePartitionInfo] =
+    clippedBlocks.iterator.buffered
+
+  private def addPartitionValues(
+      batch: Option[ColumnarBatch],
+      inPartitionValues: InternalRow): Option[ColumnarBatch] = {
+    batch.map { cb =>
+      val partitionValues = inPartitionValues.toSeq(partitionSchema)
+      val partitionScalarTypes = partitionSchema.fields.map(_.dataType)
+      val partitionScalars = partitionValues.zip(partitionScalarTypes).map {
```

Comment:
I think we should be using `safeMap` here per prior comment.  Would be nice to commonize this code.